orttraining/orttraining/training_ops/cpu/rnn/lstm_grad_compute.cc

```diff
@@ -0,0 +1,516 @@
+                                       : nullptr;
+  float* grad_Wbf = grad_bias_required ? grad_Wbo + hidden_size_
+                                       : nullptr;
+  float* grad_Wbc = grad_bias_required ? grad_Wbf + hidden_size_
+                                       : nullptr;
+  float* grad_Rbi = grad_bias_required ? grad_Wbc + hidden_size_
+                                       : nullptr;
+  float* grad_Rbo = grad_bias_required ? grad_Rbi + hidden_size_
+                                       : nullptr;
+  float* grad_Rbf = grad_bias_required ? grad_Rbo + hidden_size_
+                                       : nullptr;
+  float* grad_Rbc = grad_bias_required ? grad_Rbf + hidden_size_
+                                       : nullptr;
+
+  // Fill grad peepholes with 0s since they are used as accumulators
+  auto& grad_P = outputs.grad_peephole_weights;
+  if (grad_peephole_weights_required) {
+    std::fill_n(grad_P.data(), grad_P.size(), static_cast<T>(0));
+  }
+  float* grad_pi = grad_peephole_weights_required ? SafeRawPointer<T>(grad_P.begin(), grad_P.end(), hidden_size_x3)
+                                                  : nullptr;
+  float* grad_po = grad_peephole_weights_required ? grad_pi + hidden_size_
+                                                  : nullptr;
+  float* grad_pf = grad_peephole_weights_required ? grad_po + hidden_size_
+                                                  : nullptr;
+
+  constexpr float alpha = 1.0f;
+  // Gemm accumulation results in incorrect values. For now, use custom accumulation logic.
+  constexpr float weight_beta = 0.0f;
+
+  float* grad_ai = SafeRawPointer<T>(grad_a_span_.begin(), grad_a_span_.end(), hidden_size_x4);
+  float* grad_ao = grad_ai + hidden_size_;
+  float* grad_af = grad_ao + hidden_size_;
+  float* grad_ac = grad_af + hidden_size_;
+
+  float* grad_Wi_local = SafeRawPointer<T>(grad_W_span_.begin(), grad_W_span_.end(), weight_size);
+  float* grad_Wo_local = grad_Wi_local + hidden_sizexinput_size;
+  float* grad_Wf_local = grad_Wo_local + hidden_sizexinput_size;
+  float* grad_Wc_local = grad_Wf_local + hidden_sizexinput_size;
+
+  float* grad_Ri_local = SafeRawPointer<T>(grad_R_span_.begin(), grad_R_span_.end(), recurrence_weight_size);
+  float* grad_Ro_local = grad_Ri_local + hidden_sizexhidden_size;
+  float* grad_Rf_local = grad_Ro_local + hidden_sizexhidden_size;
+  float* grad_Rc_local = grad_Rf_local + hidden_sizexhidden_size;
+
+  float* grad_Ct2 = SafeRawPointer<T>(grad_Ct2_span_.begin(), grad_Ct2_span_.end(), hidden_size_);
+
+  std::fill_n(outputs.grad_initial_cell_state.data(), outputs.grad_initial_cell_state.size(), static_cast<T>(0));
+  std::fill_n(outputs.grad_initial_hidden_state.data(), outputs.grad_initial_hidden_state.size(), static_cast<T>(0));
+
+  for (int idx = 0; idx < batch_size_; ++idx) {
+    const size_t hidden_sizexidx = static_cast<size_t>(idx) * static_cast<size_t>(hidden_size_);
+    // maybe the first input is used and not the last.
```

Comment:
did not understand what the comment says.

Reply:
I think the comment is not in the right place. Let me update this,